compiler/rustc_metadata/src/rmeta/decoder.rs

```diff
@@ -1437,13 +1439,8 @@ impl<'a, 'tcx> CrateMetadataRef<'a> {
         })
     }
 
-    fn get_native_libraries(self, sess: &Session) -> Vec<NativeLib> {
-        if self.root.is_proc_macro_crate() {
-            // Proc macro crates do not have any *target* native libraries.
-            vec![]
-        } else {
-            self.root.native_libraries.decode((self, sess)).collect()
-        }
+    fn get_native_libraries(self, sess: &'a Session) -> impl Iterator<Item = NativeLib> + 'a {
+        self.root.native_libraries.decode((self, sess))
```

Comment:
Hm, it isn't entirely obvious to me that `native_libraries` will always decode to an  empty list for `is_proc_macro_crate() == true`, though I'll admit that I didn't spend a long time spelunking the relevant code trying to figure out if there's any code elsewhere that ensures this property holds.

Is there any way it would make sense to add an assertion or something along the lines here?

Reply:
For proc macro crates encoder now encodes only a specific set of metadata fields related to proc macros, everything else is empty.
It wasn't always like that, so decoder still has many `is_proc_macro` conditions.
We should generally cleanup it and remove them. But I guess they can be replaced with asserts instead, that would be more reliable.

Reply:
Looks like we are not checking the `is_proc_macro` property in majority of cases.
This is how it looks if the asserts are added - https://github.com/petrochenkov/rust/commit/b0268410a6e6ceab64ab788b96a25a62f28dcf1c
For `Lazy<[T]>` it looks like we cannot even check whether it's empty without decoding, that's why some checks are commented out.

So I think we should just drop the `is_proc_macro` checks without adding asserts, and assume that encoder encodes everything correctly.

Reply:
> For Lazy<[T]> it looks like we cannot even check whether it's empty without decoding, that's why some checks are commented out.

I think you can check the `.meta` field on a `Lazy<[T]>`.

Reply:
I see.
IMO, it's still simpler to assume that we are just decoding whatever the encoder put there, and make the decoder "fast and dumb", I've updated the PR description.

Reply:
I'm fine with not adding any of the asserts too, if we are sufficiently confident lack of them will not really cause any issues. Just wanted to make sure we at least considered it.